roofit/roofitcore/src/RooDataHist.cxx

```diff
@@ -1144,27 +1144,140 @@ RooPlot *RooDataHist::plotOn(RooPlot *frame, PlotOpt o) const
 
 
 ////////////////////////////////////////////////////////////////////////////////
-/// A vectorized version of RooDataHist::weight for one dimensional histograms 
-/// with no interpolation. 
-/// \param[out] output An array of weights corresponding the values in xVals.
-/// \param[in] xVals An array of coordinates for which the weights should be
+/// A vectorized version of interpolateDim for boundary safe linear 
+/// interpolation of one dimensional histograms. 
+///
+/// \param[out] output An array of interpolated weights corresponding to the 
+///                    values in xVals.
+/// \param[in] xVals An array of event coordinates for which the weights should be
 ///                  calculated.
 /// \param[in] correctForBinSize Enable the inverse bin volume correction factor.
+/// \param[in] cdfBoundaries Enable the special boundary coundition for a cdf:
+///                          Underflow bins are assumed to have weight zero and
+///                          overflow bins have weight one. Otherwise, the
+///                          histogram is mirrored at the boundaries for the
+///                          interpolation.
 
-void RooDataHist::weights(double* output, RooSpan<double const> xVals, bool correctForBinSize) 
+void RooDataHist::interpolateLinear(double* output, RooSpan<const double> xVals, 
+                                    bool correctForBinSize, bool cdfBoundaries)
 {
-  auto const nEvents = xVals.size();
-  RooAbsBinning const& binning = *_lvbins[0];
+  std::size_t const nBins = numEntries();
+  std::size_t const nEvents = xVals.size();
 
+  RooAbsBinning const& binning = *_lvbins[0];
   // Reuse the output buffer for bin indices and zero-initialize it
   auto binIndices = reinterpret_cast<int*>(output + nEvents) - nEvents; 
   std::fill(binIndices, binIndices + nEvents, 0);
-
   binning.binNumbers(xVals.data(), binIndices, nEvents);
 
-  for (std::size_t i=0; i < nEvents; ++i) {
-    auto binIdx = binIndices[i];
-    output[i] = correctForBinSize ? _wgt[binIdx] / _binv[binIdx] : _wgt[binIdx];
+  // Find coordinates of all bin centers
+  std::vector<double> binCoords(nBins);
```

Comment:
```suggestion
  // Find coordinates of all bin centers. Extend coordinates and weights with
  // one extra point before the first bin and one extra point after the last
  // bin. This means the original histogram bins span elements 1 to nBins in
  // coordsExt and weightsExt.
  std::vector<double> coordsExt(nBins+2);
  double * binCoords = coordsExt.data() + 1;
```

To avoid the copying later, why not define `binCoords` as an array inside `coordsExt` to begin with?

Reply:
I thought about this, but how would this work if we want to create e.g. `RooAbsBinning::binCenters()` in the future? But you are right, as the function looks now it is better to remove that allocation.

Reply:
It would work just the same. Like `binNumbers()`, the `binCenters()` would take an output array parameter, and in this case it would not take the beginning of the vector but `coordsExt.data() + 1`.

Reply:
That makes sense, thank you!